mjpc/planners/model_derivatives.cc

```diff
@@ -48,40 +48,117 @@ void ModelDerivatives::Compute(const mjModel* m,
-          // Jacobians
-          mjd_transitionFD(m, d, tol, mode, nullptr, nullptr,
-                           DataAt(C, t * (dim_sensor * dim_state_derivative)),
-                           nullptr);
-        } else {
-          // derivatives
-          mjd_transitionFD(
-              m, d, tol, mode,
-              DataAt(A, t * (dim_state_derivative * dim_state_derivative)),
-              DataAt(B, t * (dim_state_derivative * dim_action)),
-              DataAt(C, t * (dim_sensor * dim_state_derivative)),
-              DataAt(D, t * (dim_sensor * dim_action)));
-        }
-      });
+  // evaluate indices
+  int s = skip + 1;
+  evaluate_.push_back(0);
+  for (int t = s; t < T - s; t += s) {
+    evaluate_.push_back(t);
+  }
+  evaluate_.push_back(T - 2);
+  evaluate_.push_back(T - 1);
+
+  // interpolate indices
+  for (int t = 0; t < T; t++) {
+    if (std::find(evaluate_.begin(), evaluate_.end(), t) == evaluate_.end()) {
+      interpolate_.push_back(t);
     }
-    pool.WaitCount(count_before + T);
   }
+
+  // evaluate derivatives
+  int count_before = pool.GetCount();
+  for (int t : evaluate_) {
+    pool.Schedule([&m, &data, &A = A, &B = B, &C = C, &D = D, &x, &u, &h,
+                   dim_state, dim_state_derivative, dim_action, dim_sensor, tol,
+                   mode, t, T]() {
+      mjData* d = data[ThreadPool::WorkerId()].get();
+      // set state
+      SetState(m, d, x + t * dim_state);
+      d->time = h[t];
+
+      // set action
+      mju_copy(d->ctrl, u + t * dim_action, dim_action);
+
+      // Jacobians
+      if (t == T - 1) {
+        // Jacobians
+        mjd_transitionFD(m, d, tol, mode, nullptr, nullptr,
+                         DataAt(C, t * (dim_sensor * dim_state_derivative)),
+                         nullptr);
+      } else {
+        // derivatives
+        mjd_transitionFD(
+            m, d, tol, mode,
+            DataAt(A, t * (dim_state_derivative * dim_state_derivative)),
+            DataAt(B, t * (dim_state_derivative * dim_action)),
+            DataAt(C, t * (dim_sensor * dim_state_derivative)),
+            DataAt(D, t * (dim_sensor * dim_action)));
+      }
+    });
+  }
+  pool.WaitCount(count_before + evaluate_.size());
+  pool.ResetCount();
+
+  // interpolate derivatives
+  count_before = pool.GetCount();
+  for (int t : interpolate_) {
+    pool.Schedule([&A = A, &B = B, &C = C, &D = D, &evaluate_ = this->evaluate_,
+                   dim_state_derivative, dim_action, dim_sensor, t]() {
+      // find interval
+      int bounds[2];
+      FindInterval(bounds, evaluate_, t, evaluate_.size());
+      int e0 = evaluate_[bounds[0]];
+      int e1 = evaluate_[bounds[1]];
+
+      // normalized input
+      double tt = double(t - e0) / double(e1 - e0);
+      if (bounds[0] == bounds[1]) {
+        tt = 0.0;
+      }
+
+      // A
+      int nA = dim_state_derivative * dim_state_derivative;
+      double* Ai = DataAt(A, t * nA);
+      double* AL = DataAt(A, e0 * nA);
```

Comment:
Nit: Can you make AL and and AU `const double*` to distinguish from Ai?

Same for BL and BU.

Reply:
Done.